[Feat] 관리자는 유지관리 게시글을 작성,수정,삭제 할 수 있다

Galmaeki opened the pull request:
# Pull Request 요약

- 관리자 화면의 유지관리 게시글 작성, 수정, 삭제 기능에 대한 PR입니다

## 변경 사항

- Enums : 항목별로 줄 바꿈, 내용 주석으로 작성
- ManagerManageArticleController : 게시글 작성, 수정, 삭제 기능 추가
  - 관련 DTO 레코드 클래스 추가
- ManagerManageArticleService : 게시글 작성, 수정, 삭제 기능 추가
  - 예외처리 로직에 대해서 어떤 예외가 발생하는지 한 눈에 알 수 있도록 추출하지 않음
- ManageArticle : 
  - 일부 필드 Setter 작성
  - 어색한 컬럼명 수정
  - Office 테이블과 단방향 연결
  - delete 구현

## 관련 이슈

- #104 

## 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [x] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (2024-05-25)
- 작업 종료일: (2024-05-25)

Galmaeki committed:
[fix] Enum 줄 바꿈, 주석, 누락분 추가

문제 사항 :
- 주석 없이 상수만 존재하여 알아보기 힘든 단점이 있음

변경 사항 :
- ArticleCategory, MemberRole, Progress
  -  주석 작성 & 항목별 줄 바꿈
- ManagerPeriod
  - 누락분 추가

Galmaeki committed:
[feat] ManagerArticle Create 추가

문제 사항 :
- 주석 없이 상수만 존재하여 알아보기 힘든 단점이 있음

변경 사항 :
- ManagerManageArticleController 추가
  - create 메소드 작성
- ManagerManageArticleService 추가
  - create, createArticle 메소드 작성
- CreateManageArticleRequestDto 추가

Galmaeki committed:
[feat] ManagerArticle 필드 수정

문제 사항 :
- 일부 어색한 부분 수정

변경 사항 :
- progress 기본 값 삭제
- manager -> responsibility
- int -> Integer
- office 연관관계 추가

Galmaeki committed:
[fix] 누락 항목 추가

문제 사항 :
- 일부 필드 누락

변경 사항 :
- createArticle 메소드 누락분 추가

Galmaeki committed:
[feat] ManagerArticle update 추가

변경 사항 :
- ManageArticle 일부 필드 Setter 추가
- ManagerManageArticleController update 메소드 추가
- ManagerManageArticleService update, updateArticle 메소드 추가
- UpdateManageArticleRequestDto 생성

Galmaeki committed:
[fix] ManagerArticle update 예외처리 누락 추가

문제 사항 :
- 삭제된 게시글 확인 수정하지 않도록 변경

변경 사항 :
- 삭제된 게시글 확인 예외처리 추가

Galmaeki committed:
[feat] ManagerArticle delete 추가

변경 사항 :
- ManagerManageArticleController delete 추가
- ManagerManageArticleService delete 추가
  - 명시적 save 삭제

asd42270 commented on src/main/java/com/final_10aeat/domain/manageArticle/dto/request/CreateManageArticleRequestDto.java:
```diff
@@ -0,0 +1,19 @@
+    ManagePeriod period,
+    @NotNull
+    Integer periodCount,
+    @NotEmpty
```
엇 혹시 @ NotEmpty로 하신 이유 알 수 있을까요???

Galmaeki replied:
단순 실수였습니다! 지적해주셔서 감사합니다!

asd42270 reviewed (commented):
고생하셨습니다!!! 궁금한 점 있어서 코멘트 남겼습니다!! 하핫

Galmaeki committed:
[fix] NotEmpty -> NotBlank로 변경

문제 사항 :
- 어노테이션 사용 실수

변경 사항 :
-@NotEmpty -> @NotBlank

Galmaeki committed:
[fix] ManagerManageArticleController 접근 권한 설정

문제 사항 :
- Manager 전용 경로 권한 설정 누락

변경 사항 :
- ManagerManageArticleController 권한 설정

asd42270 reviewed (approved):
고생하셨습니다!!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/entity/ManageArticle.java:
```diff
@@ -33,32 +38,48 @@ public class ManageArticle extends SoftDeletableBaseTimeEntity {
+    @Setter
     @Column(name = "period_count")
-    private int periodCount;
+    private Integer periodCount;
```
integer로 변경하신 이유가 궁금합니다!

Galmaeki replied:
Null 허용 유무때문에 참조형을 사용하는것을 선호합니다!
지금 당장 로직에서 Null이 들어갈 가능성이 없다고 하더라도 추후 리팩토링중에 어떤 사이드 이펙트가 발생할 지 모르니 DB에 입력되는 클래스는 참조형으로 사용했습니다!

ghrltjdtprbs replied:
이 부분은 오히려 null 값이 들어가면 큰일나는 부분이라 기본형으로 변경을 생각해보시는 것도 좋을 것 같습니다! 👍👍

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ManagerManageArticleService.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private void updateArticle(UpdateManageArticleRequestDto request, ManageArticle article) {
+        if (request.period() != null) {
```
멘토님이 알려주신대로 optional로 변경하는게 더 좋을 것 같습니다! 😊

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ManagerManageArticleService.java:
```diff
@@ -0,0 +1,103 @@
+
+    private ManageArticle createArticle(CreateManageArticleRequestDto request, Office office) {
+        return ManageArticle.builder()
+            .progress(Progress.PENDING)
```
progress는 매니저가 작성시 입력하는 것이 아닌 무조건 pending으로 가져가나요??

Galmaeki replied:
스케쥴 추가 삭제마다 직접 진행도를 변경하는것 보다는 자동으로 변경되는게 사용성이 더 좋다고 생각했습니다!

ghrltjdtprbs replied:
그럼 관리 상세내역 부분 개발하면서 자동 변경 로직을 구현해야겠네요!! 좋습니다 ㅎㅎ 대기, 진행중에 대한 기준만 잘 정하시면 더 사용성이 높아질 것 같습니다.👍

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ManagerManageArticleService.java:
```diff
@@ -0,0 +1,103 @@
+            .build();
+    }
+
+    public void delete(Long manageArticleId, Manager manager) {
```
삭제된 게시물을 일괄 처리하는 로직도 추가하면 좋을 것 같습니다! 😊

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ManagerManageArticleService.java:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        if (!article.getOffice().getId().equals(manager.getOffice().getId())) {
+            throw new UnauthorizedAccessException();
```
현재 UnauthorizedAccessException는 401 http status 를 출력하니 403 으로 변경하는게 좋을 것 같습니다

ghrltjdtprbs commented on src/main/java/com/final_10aeat/common/enumclass/ArticleCategory.java:
```diff
@@ -1,5 +1,8 @@
 package com.final_10aeat.common.enumclass;
 
+
```
개행....🫠

Galmaeki replied:
파일에는 개행 되어있는데 깃에서 안보이네요....ㅠ

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/entity/ManageArticle.java:
```diff
@@ -33,32 +38,48 @@ public class ManageArticle extends SoftDeletableBaseTimeEntity {
+    @Setter
     @Column
-    private String manager;
+    private String responsibility;
```
해당 칼럼은 점검 담당을 나타내는 것 아닌가요??  왜 responsibility로 바꾸셨는지 궁금합니다! 
그리고 와이어프레임을 보니 이 부분은 받을 필요가 없는 칼럼이라 그냥 magerId로 변경하는 것도 좋을 것 같습니다!

Galmaeki replied:
여기 컬럼명은 manager로 되어있었는데
담당업체명에 대한 컬럼이라 이름을 변경했습니다!
매니저와 관계있는 컬럼은 아닙니다!

ghrltjdtprbs replied:
와이어프레임에 법정관리 쪽에 담당업체가 없는데 혹시 제가 놓친 부분이 있나요??

Galmaeki replied:
![Screenshot_20240527_150532_Chrome](https://github.com/10aeat/10aeat_Backend/assets/122267336/51f1b3a0-7628-47a4-b5f3-f59d81ae4683)
이부분입니다!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ManagerManageArticleService.java:
```diff
@@ -0,0 +1,103 @@
+            throw new ArticleAlreadyDeletedException();
+        }
+
+        if (!article.getOffice().getId().equals(manager.getOffice().getId())) {
```
같은 건물 id를 가진 manager들이 관련 게시물들을 전부 수정, 삭제할 수 있는 것은 자연스럽지 못한 것 같습니다! 
해당 게시물을 작성한 manager에게만 수정, 삭제 권한을 부여하는 것은 어떠신가요??

Galmaeki replied:
오히려 건물단위로 매니저들이 일을 하게 될텐데 그렇게되면 권한체크는 개인보다는 건물에 권한을 체크하는게 더 자연스러울 것 같았습니다!
작성자만 수정삭제를 할 경우 해당 인원의 퇴사시엔 어떤 수정도 할 수 없게 돼버리니까요!

ghrltjdtprbs replied:
유지 보수 게시글은 작성자가 직접 관리하고 유지보수 사항이 종료되면 해당 게시글도 완료되기 때문에 작성자만 수정할 수 있도록 설계했습니다!
유지관리쪽은 계속 진행되는 사항이기 때문에 확실히 건물 id만 확인하는 게 더 좋을 수도 있겠네요 굿굿👍👍

ghrltjdtprbs reviewed (approved):
수고하셨습니다! 몇 가지 리뷰 남겼으니 확인 부탁드려요!! 👍👍

Galmaeki committed:
[refactor] if -> Optional 사용으로 변경

변경 사항 :
- updateArticle 메소드 내 if문 변경

Galmaeki committed:
[fix] title null 불가능하도록 변경

변경 사항 :
- title null 불가능하도록 변경

Galmaeki committed:
[feat] 소프트 딜리트 자동삭제 기능 추가

변경 사항 :
- ManageArticleRepository: deletedAt 시간기준 검색 메소드 추가
- DeleteManageArticleScheduler: 스케쥴러 추가 (다른 엔티티와 겹치지 않는 시간대에 실행되도록 설정)

Galmaeki committed:
[fix] 사용하지 않는 import 문 삭제

변경 사항 :
- 테스트시 사용된 의존성 삭제

Galmaeki committed:
[fix] 인가 관련 에러코드 401 -> 403

변경 사항 :
- 에러코드 변경

Galmaeki committed:
Merge branch 'develop' into feature/104

Galmaeki commented:
@ghrltjdtprbs 코멘트따라 수정했습니다!
확인 부탁드림다!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/common/enumclass/Progress.java:
```diff
@@ -1,5 +1,7 @@
-    COMPLETE, INPROGRESS, PENDING
+    COMPLETE, //완료
+    INPROGRESS, //진행중
+    PENDING //미진행
```
별건 아니지만 PENDING은 대기로 적어두시는게 더 좋을것같습니다! 프로젝트 화면에서도 대기라는 단어를 사용하고 있어요!!

ghrltjdtprbs reviewed (approved):
변경사항 전부 확인했습니다! 수고하셨어요 ㅎㅎ 몇 가지 추가 리뷰 달았으니 참고만 해주시고 머지하셔도 좋을 것 같습니다 😊😊

Galmaeki committed:
[docs] 주석 변경 미진행 -> 대기

변경 사항 :
- 프로젝트에 맞도록 변경
